src/ui/EditorNS/editor.cpp

```diff
@@ -268,117 +305,184 @@ namespace EditorNS
-            newMode.insert("mime", mode.value("mime").toString());
-            newMode.insert("mode", mode.value("mode").toString());
-            langs.append(newMode);
+    QVariant Editor::getLanguageVariantData(const QString& langId)
+    {
+        auto comp = QVariant::fromValue(langId);
+        for (const auto& l : m_langCache) {
+            if (l.id == langId) {
+                QVariantMap langData;
+                langData.insert("id", l.id);
+                langData.insert("mime", l.mime);
+                langData.insert("mode", l.mode);
+                langData.insert("name", l.name);
+                langData.insert("fileNames", l.fileNames);
+                langData.insert("fileExtensions", l.fileExtensions);
+                return QVariant::fromValue(langData);
+            }
         }
-        return langs;
+        return QVariant();
+    }
+
+    QVector<Editor::LanguageData> Editor::languages()
+    {
+        return m_langCache;
     }
 
-    QString Editor::getLanguage()
+    Editor::LanguageData Editor::getLanguage()
     {
-        QVariantMap data = m_jsProxy->getRawValue("language").toMap();
-        return data.value("id").toString();
+        return m_info.language;
     }
 
     void Editor::setLanguage(const QString &language)
     {
-        sendMessage("C_CMD_SET_LANGUAGE", language);
-        if (!m_customIndentationMode)
+        for (const auto& l : m_langCache) {
+            if (l.id == language) {
+                m_info.language = l;
+                break;
+            }
+        }
+        if (!m_info.content.indentMode.custom) {
             setIndentationMode(language);
+        }
+        QString mode;
+        if (m_info.language.mime.isEmpty()) {
+            mode = m_info.language.mode;
+        } else {
+            mode = m_info.language.mime;
+        }
+        sendMessage("C_CMD_SET_LANGUAGE", mode);
+        emit languageChanged(m_info.language);
     }
 
-    void Editor::setLanguageFromFileName(QString fileName)
+    void Editor::setLanguage(const Editor::LanguageData& language)
     {
-        sendMessage("C_FUN_SET_LANGUAGE_FROM_FILENAME", fileName);
+        setLanguage(language.id);
     }
 
+    void Editor::setLanguageFromFileName(const QString& fileName)
+    {
+        // Case-sensitive fileName search
+        for (auto& l : m_langCache) {
+            if (l.fileNames.contains(fileName)) {
+                setLanguage(l.id);
+                return;
+            }
+        }
+
+        // Case insensitive file extension search
+        auto pos = fileName.lastIndexOf('.');
+        if (pos != -1) {
+            QString ext = fileName.mid(pos+1);
+            for (auto& l : m_langCache) {
+                if (l.fileExtensions.contains(ext, Qt::CaseInsensitive)) {
+                    setLanguage(l.id);
+                    return;
+                }
+            }
+        }
+    }
+/*
+    void Editor::detectLanguageFromContent()
+    {
```

Comment:
I figured out a way to do this... I'm going to push momentarily with my language cache fixes.  Turns out we were copying the language structs about 800 times over between mainwindow and frmpreferences.

I "think" I got everything working.